samza-test/src/main/java/org/apache/samza/test/framework/TestRunner.java

```diff
@@ -391,4 +395,17 @@ private void deleteDirectory(String path) {
       LOG.warn("Could not delete the directory " + path);
     }
   }
+
+  /**
+   * Test Framework only supports NoOpSerde. This method ensures null key and msg serde config for input and output streams
+   * takes preference when configs are merged in {@link org.apache.samza.execution.JobPlanner#getExecutionPlan}
+   * over {@link org.apache.samza.application.descriptors.ApplicationDescriptor} generated configs
+   */
+  private void addSerdeConfigs(StreamDescriptor descriptor) {
+    String streamIdPrefix = String.format(StreamConfig.STREAM_ID_PREFIX(), descriptor.getStreamId());
+    String keySerdeConfigKey = streamIdPrefix + StreamConfig.KEY_SERDE();
+    String msgSerdeConfigKey = streamIdPrefix + StreamConfig.MSG_SERDE();
+    this.configs.put(keySerdeConfigKey, null);
```

Comment:
Is it possible to handle this use-case without overloading and using "null", e.g., by 
defining a TestFrameworkSerde class. 
"null" in code usually signifies things gone wrong. 
see https://goo.gl/eaxLTp

Reply:
I agree with https://goo.gl/eaxLTp

The issue here is NoOpSerde is not an actual value in code (would have used that) with a facory to produce it, its absence of a serde, if we handle it the code consistently as a value everywhere, that's the right fix to do. That needs bigger refactor and more testing (using NoOpSerde applies a wider use case)

This problem will go away when I do a cleanup PR and replace appDescriptors with in-memory descriptors. This is just a temporary fix, discussed with @prateekm

Reply:
1. It is documented that we only support NoOpsSerde for now. That is intentionally done so that even if users pass a valid one, it always ends up using NoOp

2. for three cases, irrelevant of api you write test in (low level api, high level api, legacy low level)
- In the client library: for example, the library for publishing to Kafka and consuming from Kafka supports pluggable serialization.
A: Does not apply to InMemory System

- In the task implementation: your process method can use raw byte arrays as inputs and outputs, and do any parsing and serialization itself.
- Between the two: Samza provides a layer of serializers and deserializers, or serdes for short.
A: All streams are configured to have NoOpSerde only internally, that means test fw works with deserialized objects only

Reply:
@Sanil15 From your response, I think we only support `NoOpSerde` in `TestRunner` API. So the serde part question doesn't matter for which API.

I think then we wouldn't have to consider all the above cases for both low-level and high-level API.  Let's document this and provide examples of using the TestRunner API when users define custom serde's.

Reply:
@Sanil15 Just curious, Do you know how does [IntermediateStreamSerde](https://github.com/apache/samza/blob/master/samza-core/src/main/java/org/apache/samza/serializers/IntermediateMessageSerde.java), [ControlMessageSerde](https://github.com/apache/samza/blob/master/samza-api/src/main/java/org/apache/samza/system/ControlMessage.java) works with  `TestRunner`, since we only support NoOpSerde with `TestRunner`. Is the whole message converted to bytes before it reaches the InmemorySystem?

Reply:
Discussed Offline

Reply:
@shanthoosh Why are intermediate stream and control message serde relevant here? They're framework internal and not used for inputs. 

Also, "test runner only supports NoOpSerde" doesn't mean much, if anything. It'll be clearer to clarify what's expected of the user - that the inputs to any in memory streams must be already deserialized.

Reply:
sure